queryable-spec.ts

```diff
@@ -0,0 +1,442 @@
+ * used to represent both available orderings (i.e. orderings that may be
+ * requested by callers) and returned orderings (i.e. orderings followed
+ * by the returned iterators).
+ */
+interface QueryOperationOrder<T extends TermName | RDF.Variable> {
+  cost: QueryOperationCost;
+  terms: { term: T, direction: 'asc' | 'desc' }[];
+}
+
+interface QueryResultCardinality {
+  /**
+   * indicates the type of counting that was done, and MUST either be 
+   * "estimate" or "exact".
+   */
+  type: 'estimate' | 'exact';
+
+  /**
+   * Indicates an estimate of the number of quads in the stream if 
+   * type = "estimate", or the exact number of quads in the stream if 
+   * type = "exact".
+   */
+  value: number;
+}
+
+/**
+ * A QueryResultMetadata is an object that contains metadata about a certain
+ * query result.
+ */
+interface QueryResultMetadata<OrderItemsType extends TermName | RDF.Variable> {
+
+  /**
+   * An optional field that contains metadata about the number of quads in the
+   * result stream.
+   */
+  cardinality?: QueryResultCardinality;
+
+  /**
+   * An optional field that contains the available options for quad sorting
+   * based on the provided pattern, expression and options.
+   */
+  availableOrders?: QueryOperationOrder<OrderItemsType>[];
+
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+}
+
+/**
+ * A QueryResultMetadataOptions is an object that gives suggestions on what
+ * type of metadata is desired, such as when invoking FilterResult.metadata.
+ */
+ interface QueryResultMetadataOptions {
+
+  /**
+   * optional field that MAY either contain "estimate" or "exact". If defined,
+   * this type MUST correspond to the type in QueryResultMetadataCardinality.
+   */
+  cardinality?: 'estimate' | 'exact';
+
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+};
+
+interface QueryResultExecuteOptions<OrderItemsType extends TermName | RDF.Variable> {
+  
+  /**
+   * TBD
+   */
+  order?: QueryOperationOrder<OrderItemsType>;
+
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+}
+
+/**
+ * Generic interface that defines the API pattern for query result objects.
+ */
+interface BaseQueryResult {
+  
+  type: string;
+
+  /**
+   * Returns either a stream containing all the items that match the given query,
+   * a boolean or void depending on the semantics of the given query.
+   */
+  execute(opts?: any): Promise<Stream<any> | boolean | void>;
+
+  /**
+   * Asynchronously metadata of the current result.
+   */
+  metadata?(opts?: QueryResultMetadataOptions): Promise<QueryResultMetadata<any>>;
+
+  /**
+   * Asynchronously returns a boolean indicating if the requested expression is
+   * supported. If it returns true, execute() and metadata() MAY produce valid 
+   * results. If it returns false, execute() MUST return a stream emitting an 
+   * error, and metadata() MUST reject.
+   */
+  isSupported(): Promise<boolean>;
```

Comment:
Sorry to have missed this before, but why does this return a promise? Surely by the time the engine returns a result, it knows whether it supports the type?

Reply:
This is to enable users to test if a query could be executed (after the result type has been identified), without requiring the query to be (fully) executed.

This is for example useful during query optimization if a query engine would like to check if it can pass a query to a source (or another query engine).

I do think we should make it clear that `isSupported` should not necessarily be called before `execute`.

Reply:
> Sorry to have missed this before, but why does this return a promise? Surely by the time the engine returns a result, it knows whether it supports the type?

Wish I'd seen this sooner as it nicely ties into the naming discussion below. From my perspective, independent and asynchronous `isSupported()`, `execute()` and `metadata()` give us support for _lazily_ running whatever code paths they need while still allowing for memoization of shared paths. They allow engines to run through as little code as possible no matter which methods are invoked and in what order.

Reply:
My problem with this is that it requires two indirections to get to an answer.
```
let query = await engine.quads('WHAT');
let supported = await query.isSupported();
```
If we're genuinely treating the "query" as a future (which I like), then the engine call could be synchronous. No?

Reply:
Mh... With all of `isSupported()`, `execute()` and `metadata()` being asynchronous, then it might make sense to make the `query()` method and the `matchExpression()` method synchronous. I think I am in favor of that. @rubensworks ?

Reply:
Oh no, I keep forgetting that determination of `type` can also be an asynchronous operation, which is (I think) the main reason behind `query()` currently being asynchronous. We could turn `type` into `type(): Promise<'quads'|'bindings'|'boolean'|'void'>` thus making things a bit more lazy, which should be a good thing.

Reply:
I don't think there's a way around this without keeping type determination asynchronous. The only things that comes to my mind is to move `type` into the result values of both `execute()` and `metadata()`.

```
const query = engine.query();
const { type } = await query.metadata();
const { type, stream } = await query.execute();
```

This would allow for a single asynchronous indirection in the case of hitting `execute()` directly:

```
const { type, stream } = await engine.query().execute();
```

It would also still support type determination without (full) query execution:

```
const query = engine.query();
const { type } = query.metadata();

switch (type) {
    case 'bindings': /* ... */ break;
}
```

It would also tie in nicely with higher-level methods:

```
// type is guaranteed to be `bindings`
const { type, stream } = engine.select().execute(); 
```

Reply:
The approach I proposed is not supported by TypeScript, unfortunately, as there's no way to lead the compiler to infer the type of `execute()`'s return value based on the type of `await metadata()` and vice-versa. Or, if there is, I haven't been able to find it. User-defined async type guards, which would be a simpler feature as far as type resolution goes, are also still TBD (https://github.com/microsoft/TypeScript/issues/37681).

However, I think @gsvarovsky's point reveals a potential contradiction that's worth thinking about. If type determination is synchronous, we don't need the double indirection. If type determination is asynchronous, as the current consensus says, why should an engine be forced to determine the result type of a query before at least one between `isSupported()`, `execute()` and `metadata()` is called? This is where the difference between the generic method and the type-specific (higher-level) methods might help us.

The higher-level methods, given that we can assume result types, do not need the double async indirection:

```
const stream = await engine.bindings(`SELECT ...`).execute();
```

Note that nothing would prevent an engine from satisfying a request for a non-standard result type:

```
const stream = await engine.bindings(`DESCRIBE...`).execute();
```

The only reason that I can think of in favor of having the double indirection in the higher-level methods is to make it more straightforward for these to wrap the generic `query()` method. However, I would be happy with having my life made a little harder in favor of better usability client-side.

If we were to sort out the double indirection issue on the side of the higher-level methods, the generic `query()` method could be pushed even further towards maximum flexibility or simply kept as-is, although forcing type determination a-priori doesn't sound ideal to me (but no deal breaker).

As an aside, the more I think about it and the more I wonder whether we have it vice-versa. The generic method, being further away from the SPARQL spec, might actually be the higher-level one. It might actually be easier to have the spec focus on `bindings()`, `quads()`, ... and have engines using generic methods such as `query()` (like Comunica) figure out which one to use during query evaluation.

I propose to:

1. drop double async indirection in `bindings()`, `quads()`, ...

I'm thinking about proposing to:

1. drop the `query()` method from the spec

Reply:
> drop double async indirection in bindings(), quads(), ...

👍 

But do we even need the `execute` method in those cases? Since those methods are there to simplify access as opposed to `query`, perhaps we can just make it always return the direct result. For those few cases where the user wants the metadata for instance, they should then just call `query`, and deal with the double async indirection.

So we could have something like this:

```js
const stream = await engine.bindings(`SELECT ...`);
```

> drop the query() method from the spec

👎 

For me, that's the most important part of the query spec, so I really don't want to see this go 😅

Reply:
> But do we even need the execute method in those cases? Since those methods are there to simplify access as opposed to query, perhaps we can just make it always return the direct result. 

I think it's worth keeping `execute()` and `metadata()` if only for the ordering features.

> For me, that's the most important part of the query spec, so I really don't want to see this go 😅

I myself was only thinking about it. Any comment on the following?

> If type determination is asynchronous, as the current consensus says, why should an engine be forced to determine the result type of a query before at least one between isSupported(), execute() and metadata() is called?

Regardless of whether there's a way to do the above in a TS-friendly manner, I am curious to hear your thoughts.

Reply:
> > But do we even need the execute method in those cases? Since those methods are there to simplify access as opposed to query, perhaps we can just make it always return the direct result.
> I think it's worth keeping execute() and metadata() if only for the ordering features.

Is it though? Not sure if end-users (so not query engines using it for optimizations) will really use it that often.
And for those advanced use cases where they are needed, they can just use the more verbose `query` method with async indirection.

> I myself was only thinking about it. Any comment on the following?
>> If type determination is asynchronous, as the current consensus says, why should an engine be forced to determine the result type of a query before at least one between isSupported(), execute() and metadata() is called?

So if I understand correctly, you would propose to move `type` into the return type of isSupported(), execute() and metadata(), which would enable the future object to be returned in a sync manner.

However, as you discovered that TS doesn't fully handle this in its type resolution, it might complicate matters for users of the API.

Another issue is the fact that `metadata()` is (currently) not exposed on void and boolean results. Of course, we could expose a dummy object for them, but it feels to me like this points to a design issue.

Reply:
> Is it though? Not sure if end-users (so not query engines using it for optimizations) will really use it that often.
And for those advanced use cases where they are needed, they can just use the more verbose query method with async indirection.

Convincing argument. I think I agree. So we'd have higher-level methods like this:

```
const stream = await engine.bindings(`SELECT ...`);
```

This works for me. Simple, straightforward. @gsvarovsky ?

> So if I understand correctly, you would propose to move type into the return type of isSupported(), execute() and metadata(), which would enable the future object to be returned in a sync manner.

It started like this, yes, but now I'm asking whether requiring type determination before everything else could itself be a design issue, regardless of double async indirection (which I don't like but I also don't consider a blocker). If we are to support async type determination, which we should support, I think it's worth considering how we could decouple it from the determination of other kinds of metadata. I've also thought about your comment on optional properties / methods often being a pain, which I fully agree with, and how that applies to the optional properties in our metadata objects.

Yesterday I encountered a very interesting pattern to get around the lack of user-defined async type guards which, in our case, would roughly apply as follows:
 
```
interface BaseQuery<ResultType, StreamItemsType, OrderItemsType extends TermName | RDF.Variable> {
  typed(): Promise<TypedQuery<ResultType, StreamItemsType, OrderItemsType>>;
  execute(): Promise<null | Stream<StreamItemsType>>;
  availableOrders(): Promise<null | QueryOperationOrder<OrderItemsType>[]>;
}

interface TypedQuery<ResultType, StreamItemsType, OrderItemsType extends TermName | RDF.Variable> extends BaseQuery<ResultType, StreamItemsType, OrderItemsType> {
  isBindingsQuery(): this is BindingsQuery;
  isQuadsQuery(): this is QuadsQuery;
}

interface QuadsQuery extends TypedQuery<'quads', RDF.Quad, TermName> {}
interface BindingsQuery extends TypedQuery<'bindings', Bindings, RDF.Variable> {}

type Query = QuadsQuery | BindingsQuery;
```

```
const query = <Query><unknown>{};
const typed = await query.typed();

if (typed.isBindingsQuery()) {
  // correctly inferred to be of type Stream<Bindings>
  const stream = await typed.execute();
} else if (typed.isQuadsQuery()) {
  // correctly inferred to be of type Stream<RDF.Quad>
  const stream = await typed.execute();
}
```

PROS:

* gets rid of double async indirection
* supports full type inference
* fully embraces lazy loading by decoupling all metadata from each other
* gets rid of optional properties and methods by a) decoupling metadata and b) using special return values to indicate whether something is not supported
* easy to use

CONS:

* more verbose (which I don't mind)
* doesn't support `switch()` statements on query types (which I also not mind)

Reply:
That's an interesting pattern, I have to remember that one :-)

Unless I'm misunderstanding something, I don't think it gets rid of double async indirection, but it just moves it to a different place.

What we had before: (2 awaits)
```js
const engine = { ... };
const query = await engine.query();

if (query.type === 'bindings') {
  const stream = await query.execute();
}
```

What we have with this new pattern: (2 awaits)
```js
const engine = { ... };
const typed = await engine.query().typed();

if (typed.isBindingsQuery()) {
  const stream = await typed.execute();
}
```

-----

Still not convinced that the double async indirection is a big issue though :-)
I think it's an acceptable cost for a more expressive interface. (especially since we also provide a simpler one)

Reply:
You are right, I misspoke. It does not completely remove the double async indirection but moves it around so that if you're only interested in one thing, be it the type of the stream or the available orders or whether a query is supported or not, you're always faced with just one layer of async indirection. This is valuable, IMHO, as there would be no coupling of operations imposed at the interface level.

In other words, whether an engine needs to also determine result type before being able to determine cardinality, particularly estimated cardinality (the fast one used for planning), is an implementation-specific detail. Same with whether a query is supported or not. However, as it is, the current design forces the engine to do type determination before anything else which might not necessarily require type determination.

Reply:
> but moves it around so that if you're only interested in one thing, be it the type of the stream or the available orders or whether a query is supported or not, you're always faced with just one layer of async indirection.

But even then, you still need to check what the type is before you can handle the results, right?
For example, if you want to execute the query, you somehow need to do something like this:
```js
const engine = { ... };
const someResult = await engine.query().execute();

if (someResult is of type bindings) {
  const stream = someResult;
}
```

Can this type checking (in the if-statement) happen in a sync manner?

I may be misunderstanding things though...

> However, as it is, the current design forces the engine to do type determination before anything else which might not necessarily require type determination.

AFAIK, in practise this _is_ the very first thing that a query engine will do, before it will be able to do anything else.

So all of our future object members will in some way depend on the result type:

* `execute`: requires knowledge of the type due to the return type of the results
* `metadata`: requires knowledge of the type, and also its existence depends on the type
* `isSupported`: will require internal determination of the type, before it is able to determine whether or not the expression is supported

Reply:
> So all of our future object members will in some way depend on the result type:

I'm not sure. As we're talking about asynchronous determination of the result type, we can't assume a result type for any given query type and we need to run some asynchronous logic to get to the answer. Off of the top of my head:

-  `isSupported` should not require async result type determination if the `queryFormat` property of the context does not match a format that the engine can work with (i.e. SPARQL vs  GraphQL).
- `isSupported` should not require async result type determination if the query itself is not supported (i.e. a read-only engine being asked to `UPDATE` or `DELETE`, regardless of what the final result type would be).

Reply:
> isSupported should not require async result type determination if the queryFormat property of the context does not match a format that the engine can work with (i.e. SPARQL vs GraphQL).

Comunica currently checks this asynchronously via its actor-bus system (which is fully async).
If we want to make this sync, then it would come at a great cost of reduced expressivity.

> isSupported should not require async result type determination if the query itself is not supported (i.e. a read-only engine being asked to UPDATE or DELETE, regardless of what the final result type would be).

Comunica also does this async via its actor-bus system. (query parsers may be async, so you may only know if its UPDATE or DELETE asynchronously)

Reply:
Oh no, I'm not arguing for making `isSupported()` synchronous. I agree that would result in a dramatic loss of expressivity. I am arguing against forcing async result type determination before anything else can be done. For example, with the current design:

```
const query = await engine.query('QUERY STR', { queryFormat: { language: 'someUnsupportedLanguage' } });
await query.isSupported(); // false
query.type; // ?
```

In this case, `query.isSupported()` would rightfully return `Promise<false>`, thus asynchronously telling us that the query is not supported. Good! But what would be the value of `query.type` in this case? We'd either have to make `type` optional or to have a "special" type dedicated to unsupported queries, which IMHO confirms that there's a design issue somewhere. Contrast with the pattern above:

```
const query = engine.query('QUERY STR', { queryFormat: { language: 'someUnsupportedLanguage' } });
await query.isSupported(); // false
(await query.typed()).isBindingsQuery(); // false
```

In this case, `query.isSupported()` would again rightfully return `Promise<false>` but we would have no issue when it comes to the type as that is dealt with separately _at the interface level_. The specific implementation would of course be free to cache the same object internally and always refer to that.

Reply:
Ok, I agree that something is not right there.

But to take a step back, do we even need `isSupported()`?

For reference, Comunica rejects its promise when invoking `query` if the query is not supported, or for things like syntax errors.

If I remember correctly, we introduced this at https://rdf.js.org/query-spec/ for allowing async support checking, even though the return type (and result stream) was sync. But since our requirements have changed, perhaps a promise rejection is all we need here?

Reply:
Good point. Some thoughts:

1) I think dropping `isSupported()` from the current design in favor of rejecting is a good idea. Something needs to be done to address the inconsistency above and this is a good solution as it simplifies the API.

2) I'm still unconvinced about the more general issue of coupling things that shouldn't be coupled. While experimenting with adding support for `Filterable` to quadstore, I realized that having both available orders and cardinality estimation returned as optional properties of `metadata()` makes for some awkward code both implementation wise (looking at options, choosing what to include in the result) and client wise (testing which properties are present in the options object). However, if we drop metadata options and just return all possible metadata we loose a lot of room for optimization as, for example, Quadstore can return available orders synchronously while requiring an asynchronous call for cardinality estimation. I've come to think it'd be a lot better to have different methods, such as `availableOrders()` and `cardinality(precision: 'estimated'|'exact')`. Forcing async result type resolution in order to get to the query (future result) object seems somewhat similar in its coupling effect.

Reply:
However, if we were to drop `isSupported()` and maybe split `metadata()` into `availableOrders()`, `cardinality()` and the rest, I think I'd be ok with keeping async result type determination as it is right now.

Reply:
> I've come to think it'd be a lot better to have different methods, such as availableOrders() and cardinality(precision: 'estimated'|'exact')

I don't think this is a sustainable way forward for handling metadata. If we want to keep the metadata object open for other types in the future (which we should IMO, as we can not be exhaustive), then having separate functions for each metadata entry seems to conflict with this open world idea.

In any case, metadata should be cheap to calculate, so letting the source calculate all metadata even though the client only requires a subset should not be a problem.

> Quadstore can return available orders synchronously while requiring an asynchronous call for cardinality estimation.

Does this imply that some metadata functions should be sync and others async?

(Not all engines may be able to do this in a sync manner)

> I realized that having both available orders and cardinality estimation returned as optional properties of metadata() makes for some awkward code both implementation wise (looking at options, choosing what to include in the result) and client wise (testing which properties are present in the options object).

In my experience, this can be implemented using standard feature-detection techniques. Perhaps I can help making this code less awkward?

-----

I still think we're focusing too much on trying to avoid async logic. Since it's a core feature of JS, we shouldn't be afraid to use it IMO.

Reply:
Apologies for the delay, it's holiday here and I actually managed to take some time off :). 

> I still think we're focusing too much on trying to avoid async logic. Since it's a core feature of JS, we shouldn't be afraid to use it IMO.

I agree, and that's not what I am afraid of. After having thought about it a bit more, I just happen to prefer more specialized methods with weaker coupling at the interface level:

```
const query = engine.query('QUERY STR');
await query.resultType();
await query.cardinality();
await query.availableOrders();
```

I think the above would be significantly easier to code against _and extend_ than having metadata returned through a unified `metadata()` , all the while leaving implementations free to internally cache intermediate objects if/when applicable (thus potentially leading to only the first call to either `cardinality()` or `availableOrders()` being asynchronous in practice). 

The same preference applies to result type determination, too, although the limitations of TS would require the use of a pattern like the one I referenced above and that might be too far from the simplicity of having `query.type` immediately available as per the current design. 

When it comes to common DBMSs, query builders and ORMs, it seems to me like metadata is most often - if not always - available _only_ through specific methods rather than generic ones. I've never seen cardinality being returned together with available indexes by any of the drivers, builders and ORMs I've ever used and I don't see any advantage in deviating from this particular state of things. In the worst (and most common, IMHO) case, cardinality estimation and index retrieval are two completely separate async operations. It is true that they might not be computationally expensive operations but an interface should not force implementations to run both when clients only need one.

> Does this imply that some metadata functions should be sync and others async?

No, they should all be async _at the interface level_. I was just making an example of a specific implementation (quadstore) that can return available orders in a much faster way as no querying is needed in practice. That is not (necessarily) the case with SQL-based store, in which indexes may be dynamically added / dropped.

>  then having separate functions for each metadata entry seems to conflict with this open world idea.

This is somewhat true, although the same feature-detection logic mentioned above could be applied to testing for the presence of specific metadata-related methods. Or, coming back to your point about optional things being tricky (if not annoying) to code against, implementors could just throw `not supported` errors to be caught by clients. As we're not iterating on results at this level, additional try/catch blocks should not present issues from a performance standpoint.

All this said... Time for be to stop banging on this drum, too! Too much banging achieves nothing. We do agree on dropping `isSupported()` - I'll do that later in the day. As for my objections on coupling, I do think that the interface would be objectively better with dedicated methods for metadata but I am in a minority of one, so I concede defeat :)

Reply:
Ok, you make some good points regarding the method-based metadata.
While in Comunica we don't require them atm, it makes sense to enable this within the RDF/JS spec.

I would however suggest to still have some kind of intermediary metadata object, which then exposes these methods.
This would allow this metadata object to be reused in other places (such as internally within a query engine, or during post-processing of results without requiring the actual results).

So this could look as something like this:
```
const query = await engine.query('QUERY STR');
// I think we still need query.type, right?
const results = await query.execute();

const metadata = query.metadata;
await metadata.cardinality();
await metadata.availableOrders();
```

Reply:
> `// I think we still need query.type, right?`

Yes! Unless we stumble into other workarounds for the lack of async type resolution in TS. The one I posted above works but the resulting code lacks some immediacy IMHO. However, perhaps we should consider renaming to `resultType` to disambiguate from query type?

> I would however suggest to still have some kind of intermediary metadata object, which then exposes these methods.

Absolutely fine by me.

Reply:
> However, perhaps we should consider renaming to resultType to disambiguate from query type?

Hmm yes, this might result in some possible confusion indeed.